internal/server/v2/observation/collection.go

```diff
@@ -32,5 +40,96 @@ func FetchFromCollection(
 	childType string,
 	queryDate string,
 ) (*pbv2.ObservationResponse, error) {
-	return nil, nil
+	btData, err := stat.ReadStatCollection(
+		ctx, store.BtGroup, bigtable.BtObsCollection,
+		ancestor, childType, variables, queryDate,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result := &pbv2.ObservationResponse{
+		ObservationsByVariable: map[string]*pbv2.VariableObservation{},
+		Facets:                 map[string]*pb.StatMetadata{},
+	}
+	variablesMissingData := []string{}
+	for _, variable := range variables {
+		result.ObservationsByVariable[variable] = &pbv2.VariableObservation{
+			ObservationsByEntity: map[string]*pbv2.EntityObservation{},
+		}
+		// Create a short alias
+		obsByEntity := result.ObservationsByVariable[variable].ObservationsByEntity
+		data, ok := btData[variable]
+		if !ok || data == nil {
+			variablesMissingData = append(variablesMissingData, variable)
+			continue
+		}
+		cohorts := data.SourceCohorts
+		// Sort cohort first, so the preferred source is populated first.
+		sort.Sort(ranking.CohortByRank(cohorts))
+		for _, cohort := range cohorts {
+			facet := util.GetMetadata(cohort)
+			facetID := util.GetMetadataHash(facet)
+			result.Facets[facetID] = facet
+			for entity, val := range cohort.Val {
+				if _, ok := obsByEntity[entity]; !ok {
+					obsByEntity[entity] = &pbv2.EntityObservation{}
+				}
+				// When date is in the request, response date is the given date.
+				// Otherwise, response date is the latest date from the cache.
+				respDate := queryDate
+				if respDate == LATEST {
+					respDate = cohort.PlaceToLatestDate[entity]
+				}
+				obsByEntity[entity].OrderedFacetObservations = append(
+					obsByEntity[entity].OrderedFacetObservations,
+					&pbv2.FacetObservation{
+						FacetId: facetID,
+						Observations: []*pb.PointStat{
+							{
+								Date:  respDate,
+								Value: proto.Float64(val),
+							},
+						},
+					},
+				)
+			}
+		}
+	}
+	// Check if need to read from memory database.
+	for _, variable := range variables {
+		if store.MemDb.HasStatVar(variable) {
+			variablesMissingData = append(variablesMissingData, variable)
+		}
+	}
+	// Fetch linked places if need to read data from memdb or time series Bigtable
+	// cache.
+	var childPlaces []string
+	if len(variablesMissingData) > 0 {
+		// TODO(shifucun): use V2 API
+		childPlacesMap, err := placein.GetPlacesIn(
+			ctx, store, []string{ancestor}, childType)
+		if err != nil {
+			return nil, err
+		}
+		childPlaces = childPlacesMap[ancestor]
+	}
+	if len(variablesMissingData) > 0 {
+		moreResult, err := FetchFromSeries(
+			ctx,
+			store,
+			variablesMissingData,
+			childPlaces,
+			queryDate,
+		)
+		if err != nil {
+			return nil, err
+		}
+		for variable, variableData := range moreResult.ObservationsByVariable {
+			result.ObservationsByVariable[variable] = variableData
+		}
+		for facet := range moreResult.Facets {
+			result.Facets[facet] = moreResult.Facets[facet]
+		}
```

Comment:
nit:

for facet, res := range moreResult.Facets {
			result.Facets[facet] = res
}

Reply:
Done